File: client-stub/Protocol/Messages.cs
// Espejo en C# de server/internal/protocol/messages.go
// Debe mantenerse en sincronía manualmente con ese archivo y con /common/protocol/PROTOCOL.md
// hasta que se automatice la generación (ej. vía un generador desde un schema compartido).
//
// Pensado para usarse dentro de un proyecto Godot 4 (C#) como capa de Network/Protocol.

using System;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PokemonOnline.Protocol
{
    public class Envelope
    {
        [JsonPropertyName("type")]
        public string Type { get; set; } = "";

        [JsonPropertyName("seq")]
        public long Seq { get; set; }

        [JsonPropertyName("payload")]
        public JsonElement Payload { get; set; }
    }

    public class ErrorPayload
    {
        [JsonPropertyName("code")] public string Code { get; set; } = "";
        [JsonPropertyName("message")] public string Message { get; set; } = "";
    }

    public class LoginPayload
    {
        [JsonPropertyName("username")] public string Username { get; set; } = "";
        [JsonPropertyName("password")] public string Password { get; set; } = "";
        // Si se completa, reautentica con un JWT de una sesión anterior (login_ok.SessionToken)
        // en vez de usuario/contraseña. Útil para reconectar el WebSocket sin volver a pedirlas.
        [JsonPropertyName("session_token")] public string? SessionToken { get; set; }
    }

    public class LoginOKPayload
    {
        [JsonPropertyName("account_id")] public string AccountId { get; set; } = "";
        [JsonPropertyName("character_id")] public string CharacterId { get; set; } = "";
        [JsonPropertyName("session_token")] public string SessionToken { get; set; } = "";
        [JsonPropertyName("map_id")] public string MapId { get; set; } = "";
        [JsonPropertyName("pos_x")] public int PosX { get; set; }
        [JsonPropertyName("pos_y")] public int PosY { get; set; }
        [JsonPropertyName("color")] public string Color { get; set; } = "default";
    }

    public class MovePayload
    {
        [JsonPropertyName("map_id")] public string MapId { get; set; } = "";
        [JsonPropertyName("x")] public int X { get; set; }
        [JsonPropertyName("y")] public int Y { get; set; }
        [JsonPropertyName("facing")] public string Facing { get; set; } = "down";
        [JsonPropertyName("state")] public string State { get; set; } = "idle";
    }

    public class PlayerUpdatePayload
    {
        [JsonPropertyName("character_id")] public string CharacterId { get; set; } = "";
        [JsonPropertyName("nickname")] public string Nickname { get; set; } = "";
        [JsonPropertyName("sprite_id")] public string SpriteId { get; set; } = "";
        [JsonPropertyName("map_id")] public string MapId { get; set; } = "";
        [JsonPropertyName("x")] public int X { get; set; }
        [JsonPropertyName("y")] public int Y { get; set; }
        [JsonPropertyName("facing")] public string Facing { get; set; } = "";
        [JsonPropertyName("state")] public string State { get; set; } = "";
        [JsonPropertyName("color")] public string Color { get; set; } = "default";
    }

    public class SetColorPayload
    {
        [JsonPropertyName("color")] public string Color { get; set; } = "default";
    }

    // Solo al emisor, cuando su "move" excede la velocidad físicamente posible: X/Y son la
    // última posición válida conocida por el servidor, para resincronizarse.
    public class MoveRejectedPayload
    {
        [JsonPropertyName("map_id")] public string MapId { get; set; } = "";
        [JsonPropertyName("x")] public int X { get; set; }
        [JsonPropertyName("y")] public int Y { get; set; }
        [JsonPropertyName("facing")] public string Facing { get; set; } = "";
    }

    // Una sola vez, justo tras login_ok: quién ya está en el mapa de spawn.
    public class MapPlayersSnapshotPayload
    {
        [JsonPropertyName("players")] public PlayerUpdatePayload[] Players { get; set; } = Array.Empty<PlayerUpdatePayload>();
    }

    public class SendChatPayload
    {
        [JsonPropertyName("channel")] public string Channel { get; set; } = "local";
        [JsonPropertyName("target_character_id")] public string? TargetCharacterId { get; set; }
        [JsonPropertyName("message")] public string Message { get; set; } = "";
    }

    public class ChatMessagePayload
    {
        [JsonPropertyName("channel")] public string Channel { get; set; } = "";
        [JsonPropertyName("from_character_id")] public string FromCharacterId { get; set; } = "";
        [JsonPropertyName("from_nickname")] public string FromNickname { get; set; } = "";
        [JsonPropertyName("message")] public string Message { get; set; } = "";
        [JsonPropertyName("timestamp")] public string Timestamp { get; set; } = "";
    }

    // ---- Pokémon (resumen para UI) ----

    public class PokemonSummaryPayload
    {
        [JsonPropertyName("id")] public string Id { get; set; } = "";
        [JsonPropertyName("species_id")] public int SpeciesId { get; set; }
        [JsonPropertyName("nickname")] public string? Nickname { get; set; }
        [JsonPropertyName("level")] public int Level { get; set; }
        [JsonPropertyName("location")] public string Location { get; set; } = "";
    }

    public class MyPokemonListPayload
    {
        [JsonPropertyName("pokemon")] public PokemonSummaryPayload[] Pokemon { get; set; } = Array.Empty<PokemonSummaryPayload>();
    }

    // ---- Trade ----

    public class TradeRequestPayload
    {
        [JsonPropertyName("target_character_id")] public string TargetCharacterId { get; set; } = "";
    }

    public class TradeRequestReceivedPayload
    {
        [JsonPropertyName("trade_session_id")] public string TradeSessionId { get; set; } = "";
        [JsonPropertyName("from_character_id")] public string FromCharacterId { get; set; } = "";
        [JsonPropertyName("from_nickname")] public string FromNickname { get; set; } = "";
    }

    public class TradeOfferUpdatedPayload
    {
        [JsonPropertyName("trade_session_id")] public string TradeSessionId { get; set; } = "";
        [JsonPropertyName("character_id")] public string CharacterId { get; set; } = "";
        [JsonPropertyName("pokemon")] public PokemonSummaryPayload Pokemon { get; set; } = new();
    }

    public class TradeSessionRefPayload
    {
        [JsonPropertyName("trade_session_id")] public string TradeSessionId { get; set; } = "";
    }

    public class TradeOfferSetPayload
    {
        [JsonPropertyName("trade_session_id")] public string TradeSessionId { get; set; } = "";
        [JsonPropertyName("pokemon_id")] public string PokemonId { get; set; } = "";
    }

    public class TradeCompletedPayload
    {
        [JsonPropertyName("trade_session_id")] public string TradeSessionId { get; set; } = "";
        [JsonPropertyName("char_a_received_pokemon_id")] public string CharAReceivedId { get; set; } = "";
        [JsonPropertyName("char_b_received_pokemon_id")] public string CharBReceivedId { get; set; } = "";
    }

    public class TradeCancelledPayload
    {
        [JsonPropertyName("trade_session_id")] public string TradeSessionId { get; set; } = "";
        [JsonPropertyName("reason")] public string Reason { get; set; } = "";
    }

    // ---- Amigos ----

    public class FriendRequestPayload
    {
        [JsonPropertyName("target_username")] public string TargetUsername { get; set; } = "";
    }

    public class FriendRequestReceivedPayload
    {
        [JsonPropertyName("from_account_id")] public string FromAccountId { get; set; } = "";
        [JsonPropertyName("from_username")] public string FromUsername { get; set; } = "";
    }

    public class FriendRefPayload
    {
        [JsonPropertyName("target_account_id")] public string TargetAccountId { get; set; } = "";
    }

    public class FriendListEntryPayload
    {
        [JsonPropertyName("account_id")] public string AccountId { get; set; } = "";
        [JsonPropertyName("username")] public string Username { get; set; } = "";
        [JsonPropertyName("online")] public bool Online { get; set; }
    }

    public class FriendListPayload
    {
        [JsonPropertyName("friends")] public FriendListEntryPayload[] Friends { get; set; } = Array.Empty<FriendListEntryPayload>();
    }

    public class FriendStatusUpdatePayload
    {
        [JsonPropertyName("account_id")] public string AccountId { get; set; } = "";
        [JsonPropertyName("online")] public bool Online { get; set; }
    }

    // ---- Grupos (party) ----

    public class PartyInvitePayload
    {
        [JsonPropertyName("target_character_id")] public string TargetCharacterId { get; set; } = "";
    }

    public class PartyInviteReceivedPayload
    {
        [JsonPropertyName("party_id")] public string PartyId { get; set; } = "";
        [JsonPropertyName("from_character_id")] public string FromCharacterId { get; set; } = "";
        [JsonPropertyName("from_nickname")] public string FromNickname { get; set; } = "";
    }

    public class PartyRefPayload
    {
        [JsonPropertyName("party_id")] public string PartyId { get; set; } = "";
    }

    public class PartyMemberPayload
    {
        [JsonPropertyName("character_id")] public string CharacterId { get; set; } = "";
        [JsonPropertyName("nickname")] public string Nickname { get; set; } = "";
        [JsonPropertyName("map_id")] public string MapId { get; set; } = "";
        [JsonPropertyName("x")] public int X { get; set; }
        [JsonPropertyName("y")] public int Y { get; set; }
        [JsonPropertyName("is_leader")] public bool IsLeader { get; set; }
    }

    public class PartyUpdatePayload
    {
        [JsonPropertyName("party_id")] public string PartyId { get; set; } = "";
        [JsonPropertyName("members")] public PartyMemberPayload[] Members { get; set; } = Array.Empty<PartyMemberPayload>();
    }

    public class PartyDisbandedPayload
    {
        [JsonPropertyName("party_id")] public string PartyId { get; set; } = "";
        [JsonPropertyName("reason")] public string Reason { get; set; } = "";
    }

    // ---- Mercado (asincrónico) ----

    public class MarketListPayload
    {
        [JsonPropertyName("pokemon_id")] public string PokemonId { get; set; } = "";
        [JsonPropertyName("price")] public int Price { get; set; }
    }

    public class MarketListingRefPayload
    {
        [JsonPropertyName("listing_id")] public string ListingId { get; set; } = "";
    }

    public class MarketListingPayload
    {
        [JsonPropertyName("listing_id")] public string ListingId { get; set; } = "";
        [JsonPropertyName("seller_char_id")] public string SellerCharId { get; set; } = "";
        [JsonPropertyName("seller_nickname")] public string SellerNickname { get; set; } = "";
        [JsonPropertyName("pokemon")] public PokemonSummaryPayload Pokemon { get; set; } = new();
        [JsonPropertyName("price")] public int Price { get; set; }
    }

    public class MarketListingsPayload
    {
        [JsonPropertyName("listings")] public MarketListingPayload[] Listings { get; set; } = Array.Empty<MarketListingPayload>();
    }

    public class MarketPurchasedPayload
    {
        [JsonPropertyName("listing_id")] public string ListingId { get; set; } = "";
        [JsonPropertyName("pokemon")] public PokemonSummaryPayload Pokemon { get; set; } = new();
        [JsonPropertyName("price")] public int Price { get; set; }
    }

    public class MarketSoldPayload
    {
        [JsonPropertyName("listing_id")] public string ListingId { get; set; } = "";
        [JsonPropertyName("buyer_nickname")] public string BuyerNickname { get; set; } = "";
        [JsonPropertyName("price")] public int Price { get; set; }
    }

    // ---- Gremios (persistentes, a diferencia de party) ----

    public class GuildCreatePayload
    {
        [JsonPropertyName("name")] public string Name { get; set; } = "";
    }

    public class GuildInvitePayload
    {
        [JsonPropertyName("target_character_id")] public string TargetCharacterId { get; set; } = "";
    }

    public class GuildInviteReceivedPayload
    {
        [JsonPropertyName("guild_id")] public string GuildId { get; set; } = "";
        [JsonPropertyName("guild_name")] public string GuildName { get; set; } = "";
        [JsonPropertyName("from_character_id")] public string FromCharacterId { get; set; } = "";
        [JsonPropertyName("from_nickname")] public string FromNickname { get; set; } = "";
    }

    public class GuildRefPayload
    {
        [JsonPropertyName("guild_id")] public string GuildId { get; set; } = "";
    }

    public class GuildKickPayload
    {
        [JsonPropertyName("target_character_id")] public string TargetCharacterId { get; set; } = "";
    }

    public class GuildMemberPayload
    {
        [JsonPropertyName("character_id")] public string CharacterId { get; set; } = "";
        [JsonPropertyName("nickname")] public string Nickname { get; set; } = "";
        [JsonPropertyName("online")] public bool Online { get; set; }
        [JsonPropertyName("is_officer")] public bool IsOfficer { get; set; }
    }

    public class GuildUpdatePayload
    {
        [JsonPropertyName("guild_id")] public string GuildId { get; set; } = "";
        [JsonPropertyName("name")] public string Name { get; set; } = "";
        [JsonPropertyName("members")] public GuildMemberPayload[] Members { get; set; } = Array.Empty<GuildMemberPayload>();
    }

    public class GuildDisbandedPayload
    {
        [JsonPropertyName("guild_id")] public string GuildId { get; set; } = "";
        [JsonPropertyName("reason")] public string Reason { get; set; } = "";
    }

    // Helper de (de)serialización, para no repetir JsonSerializer.Serialize/Deserialize
    // en cada punto del cliente que arma o lee un Envelope.
    public static class ProtocolCodec
    {
        public static string Encode(string type, object payload)
        {
            var payloadJson = JsonSerializer.Serialize(payload);
            var envelope = new
            {
                type,
                payload = JsonDocument.Parse(payloadJson).RootElement
            };
            return JsonSerializer.Serialize(envelope);
        }

        public static T DecodePayload<T>(Envelope envelope)
        {
            return envelope.Payload.Deserialize<T>()
                ?? throw new InvalidOperationException("payload nulo o con formato inesperado");
        }
    }
}

File: server/internal/protocol/messages.go
// Package protocol define el contrato de mensajes cliente<->servidor.
// Debe mantenerse en sincronía con /common/protocol/PROTOCOL.md y con el
// stub de cliente en /client-stub/Protocol. Cualquier cambio aquí es un
// cambio de contrato: avisar en ambos lados.
package protocol

import "encoding/json"

// Envelope es la envoltura común de todo mensaje, en ambas direcciones.
type Envelope struct {
	Type    string          `json:"type"`
	Seq     int64           `json:"seq,omitempty"`
	Payload json.RawMessage `json:"payload"`
}

// ---- Autenticación ----

type LoginPayload struct {
	Username string `json:"username,omitempty"`
	Password string `json:"password,omitempty"`
	// SessionToken, si viene presente, reautentica con el JWT emitido en un login_ok
	// previo y hace que Username/Password se ignoren. Pensado para reconexión del
	// WebSocket sin volver a pedir credenciales.
	SessionToken string `json:"session_token,omitempty"`
}

type LoginOKPayload struct {
	AccountID    string `json:"account_id"`
	CharacterID  string `json:"character_id"`
	SessionToken string `json:"session_token"`
	MapID        string `json:"map_id"`
	PosX         int    `json:"pos_x"`
	PosY         int    `json:"pos_y"`
	// Color es el color de sprite persistido del personaje ("default" si nunca lo cambió) —
	// el cliente lo necesita para saber qué mostrar como seleccionado en el picker y para
	// mandarlo él mismo en su primer "move" (ver world.AllowedSpriteColors).
	Color string `json:"color"`
}

type ErrorPayload struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// ---- Movimiento / presencia ----

type MovePayload struct {
	MapID  string `json:"map_id"`
	X      int    `json:"x"`
	Y      int    `json:"y"`
	Facing string `json:"facing"`
	State  string `json:"state"` // walking | idle | battling
}

type PlayerUpdatePayload struct {
	CharacterID string `json:"character_id"`
	Nickname    string `json:"nickname"`
	SpriteID    string `json:"sprite_id"`
	MapID       string `json:"map_id"`
	X           int    `json:"x"`
	Y           int    `json:"y"`
	Facing      string `json:"facing"`
	State       string `json:"state"`
	Color       string `json:"color"`
}

// SetColorPayload cambia el color de sprite del emisor — ver world.AllowedSpriteColors para
// los valores válidos. El servidor rechaza cualquier otro valor (no es un campo de texto libre).
type SetColorPayload struct {
	Color string `json:"color"`
}

// MoveRejectedPayload se manda de vuelta SOLO al emisor cuando su "move" excede la
// velocidad físicamente posible (ver ws.Hub.UpdatePosition) — X/Y son la posición corregida
// (la última válida conocida por el servidor) para que el cliente se resincronice.
type MoveRejectedPayload struct {
	MapID  string `json:"map_id"`
	X      int    `json:"x"`
	Y      int    `json:"y"`
	Facing string `json:"facing"`
}

// MapPlayersSnapshotPayload se manda UNA vez, justo tras login_ok, con quién ya está en el
// mapa de spawn del jugador. Sin esto, un jugador recién conectado no ve a nadie hasta que
// esa otra persona mande su primer "move" (player_update no se re-emite espontáneamente).
type MapPlayersSnapshotPayload struct {
	Players []PlayerUpdatePayload `json:"players"`
}

// ---- Chat ----

type SendChatPayload struct {
	Channel           string `json:"channel"` // global | local | private | command
	TargetCharacterID string `json:"target_character_id,omitempty"`
	Message           string `json:"message"`
}

type ChatMessagePayload struct {
	Channel         string `json:"channel"`
	FromCharacterID string `json:"from_character_id"`
	FromNickname    string `json:"from_nickname"`
	Message         string `json:"message"`
	Timestamp       string `json:"timestamp"`
}

// ---- Pokémon (resumen para UI, no el detalle completo de combate) ----

type PokemonSummaryPayload struct {
	ID        string `json:"id"`
	SpeciesID int    `json:"species_id"`
	Nickname  string `json:"nickname,omitempty"`
	Level     int    `json:"level"`
	Location  string `json:"location"` // team | pc | in_trade
}

type MyPokemonListPayload struct {
	Pokemon []PokemonSummaryPayload `json:"pokemon"`
}

// ---- Trade ----

type TradeRequestPayload struct {
	TargetCharacterID string `json:"target_character_id"`
}

// TradeRequestReceivedPayload se manda al destinatario de un trade_request. Incluye el
// nickname (no solo el character_id) porque el cliente todavía no tiene forma de resolver
// nicknames a partir de un ID salvo por los jugadores que ya vio moverse en su mapa.
type TradeRequestReceivedPayload struct {
	TradeSessionID  string `json:"trade_session_id"`
	FromCharacterID string `json:"from_character_id"`
	FromNickname    string `json:"from_nickname"`
}

// TradeOfferUpdatedPayload se manda a AMBOS participantes cada vez que uno fija su oferta,
// para que la UI de cada uno pueda mostrar en vivo qué ofreció el otro antes de confirmar
// (sin esto, cada jugador confirmaría "a ciegas", sin ver qué le están por dar a cambio).
type TradeOfferUpdatedPayload struct {
	TradeSessionID string                `json:"trade_session_id"`
	CharacterID    string                `json:"character_id"` // quién fijó esta oferta
	Pokemon        PokemonSummaryPayload `json:"pokemon"`
}

type TradeOfferSetPayload struct {
	TradeSessionID string `json:"trade_session_id"`
	PokemonID      string `json:"pokemon_id"`
}

type TradeConfirmPayload struct {
	TradeSessionID string `json:"trade_session_id"`
}

type TradeCompletedPayload struct {
	TradeSessionID  string `json:"trade_session_id"`
	CharAReceivedID string `json:"char_a_received_pokemon_id"`
	CharBReceivedID string `json:"char_b_received_pokemon_id"`
}

// TradeSessionRefPayload es el payload compartido por trade_accept y trade_decline:
// ambos solo necesitan identificar la sesión.
type TradeSessionRefPayload struct {
	TradeSessionID string `json:"trade_session_id"`
}

type TradeCancelledPayload struct {
	TradeSessionID string `json:"trade_session_id"`
	Reason         string `json:"reason"`
}

// ---- Amigos ----

type FriendRequestPayload struct {
	TargetUsername string `json:"target_username"`
}

type FriendRequestReceivedPayload struct {
	FromAccountID string `json:"from_account_id"`
	FromUsername  string `json:"from_username"`
}

// FriendRefPayload identifica a la otra cuenta en friend_accept/friend_decline/friend_remove.
type FriendRefPayload struct {
	TargetAccountID string `json:"target_account_id"`
}

type FriendListEntryPayload struct {
	AccountID string `json:"account_id"`
	Username  string `json:"username"`
	Online    bool   `json:"online"`
}

type FriendListPayload struct {
	Friends []FriendListEntryPayload `json:"friends"`
}

type FriendStatusUpdatePayload struct {
	AccountID string `json:"account_id"`
	Online    bool   `json:"online"`
}

// ---- Grupos (party) ----

type PartyInvitePayload struct {
	TargetCharacterID string `json:"target_character_id"`
}

type PartyInviteReceivedPayload struct {
	PartyID         string `json:"party_id"`
	FromCharacterID string `json:"from_character_id"`
	FromNickname    string `json:"from_nickname"`
}

// PartyRefPayload identifica al grupo en party_accept/party_decline/party_leave.
type PartyRefPayload struct {
	PartyID string `json:"party_id"`
}

type PartyMemberPayload struct {
	CharacterID string `json:"character_id"`
	Nickname    string `json:"nickname"`
	MapID       string `json:"map_id"`
	X           int    `json:"x"`
	Y           int    `json:"y"`
	IsLeader    bool   `json:"is_leader"`
}

type PartyUpdatePayload struct {
	PartyID string               `json:"party_id"`
	Members []PartyMemberPayload `json:"members"`
}

type PartyDisbandedPayload struct {
	PartyID string `json:"party_id"`
	Reason  string `json:"reason"`
}

// ---- Mercado (asincrónico — a diferencia de trade, no necesita a los dos jugadores online) ----

type MarketListPayload struct {
	PokemonID string `json:"pokemon_id"`
	Price     int    `json:"price"`
}

// MarketListingRefPayload identifica una publicación en market_cancel/market_buy y en las
// confirmaciones market_list_ok/market_cancelled.
type MarketListingRefPayload struct {
	ListingID string `json:"listing_id"`
}

type MarketListingPayload struct {
	ListingID      string                `json:"listing_id"`
	SellerCharID   string                `json:"seller_char_id"`
	SellerNickname string                `json:"seller_nickname"`
	Pokemon        PokemonSummaryPayload `json:"pokemon"`
	Price          int                   `json:"price"`
}

// MarketListingsPayload es la respuesta tanto a market_browse (publicaciones de todos) como a
// market_my_listings (solo las propias) — el cliente ya sabe cuál pidió, no hace falta
// distinguir el tipo de mensaje.
type MarketListingsPayload struct {
	Listings []MarketListingPayload `json:"listings"`
}

// MarketPurchasedPayload va al COMPRADOR: qué recibió y por cuánto.
type MarketPurchasedPayload struct {
	ListingID string                `json:"listing_id"`
	Pokemon   PokemonSummaryPayload `json:"pokemon"`
	Price     int                   `json:"price"`
}

// MarketSoldPayload va al VENDEDOR (si está conectado en ese momento): que se vendió, quién
// compró y por cuánto. Si el vendedor está offline simplemente no lo recibe — se entera la
// próxima vez que abra "mis publicaciones" (ya no va a estar listada).
type MarketSoldPayload struct {
	ListingID     string `json:"listing_id"`
	BuyerNickname string `json:"buyer_nickname"`
	Price         int    `json:"price"`
}

// ---- Gremios (persistentes — a diferencia de party, no se disuelven al desconectarse todos) ----

type GuildCreatePayload struct {
	Name string `json:"name"`
}

type GuildInvitePayload struct {
	TargetCharacterID string `json:"target_character_id"`
}

type GuildInviteReceivedPayload struct {
	GuildID         string `json:"guild_id"`
	GuildName       string `json:"guild_name"`
	FromCharacterID string `json:"from_character_id"`
	FromNickname    string `json:"from_nickname"`
}

// GuildRefPayload identifica al gremio en guild_accept/guild_decline.
type GuildRefPayload struct {
	GuildID string `json:"guild_id"`
}

type GuildKickPayload struct {
	TargetCharacterID string `json:"target_character_id"`
}

type GuildMemberPayload struct {
	CharacterID string `json:"character_id"`
	Nickname    string `json:"nickname"`
	Online      bool   `json:"online"`
	IsOfficer   bool   `json:"is_officer"`
}

type GuildUpdatePayload struct {
	GuildID string                `json:"guild_id"`
	Name    string                `json:"name"`
	Members []GuildMemberPayload `json:"members"`
}

type GuildDisbandedPayload struct {
	GuildID string `json:"guild_id"`
	Reason  string `json:"reason"`
}

// Helper para construir un Envelope de salida sin repetir json.Marshal en cada handler.
func NewEnvelope(msgType string, payload any) (Envelope, error) {
	raw, err := json.Marshal(payload)
	if err != nil {
		return Envelope{}, err
	}
	return Envelope{Type: msgType, Payload: raw}, nil
}
